File: CSharp/9_linked_list/_21_Merge_Two_Sorted_Lists.cs
using System;
using static System.Console;
using static System.DateTime;
using static System.TimeSpan;

public static class _21_Merge_Two_Sorted_Lists
{
    public class ListNode
    {
        public int val;
        public ListNode next;

        public ListNode(int val = 0, ListNode next = null)
        {
            this.val = val;
            this.next = next;
        }

        public override string ToString()
        {
            var sb = new System.Text.StringBuilder();
            sb.Append('[');
            var current = this;
            while (current != null)
            {
                sb.Append(current.val);
                if (current.next != null)
                    sb.Append(" -> ");
                current = current.next;
            }

            sb.Append(']');
            return sb.ToString();
        }
    }

    public static ListNode MergeTwoLists(ListNode l1, ListNode l2)
    {
        ListNode dummy = new(-1);
        ListNode current = dummy;

        while (l1 != null && l2 != null)
        {
            if (l1.val < l2.val)
            {
                current.next = l1;
                l1 = l1.next;
            }
            else
            {
                current.next = l2;
                l2 = l2.next;
            }

            current = current.next;
        }

        if (l1 != null)
            current.next = l1;
        else
            current.next = l2;

        return dummy.next;
    }

    public static void TestMergeTwoLists(ListNode list1, ListNode list2)
    {
        WriteLine("List 1: " + list1);
        WriteLine("List 2: " + list2);

        var start = Now.Ticks / TicksPerMillisecond;
        ListNode result = MergeTwoLists(list1, list2);
        var end = Now.Ticks / TicksPerMillisecond;

        WriteLine("Result: " + result);
        WriteLine("Execution time: " + (end - start) + "ms");
    }
}

File: C/9_linked_list/_21_Merge_Two_Sorted_Lists.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* createNode(int val) {
    struct ListNode *node = (struct ListNode *) malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

struct ListNode* mergeTwoLists(struct ListNode *list1, struct ListNode *list2) {
    struct ListNode *dummy = createNode(-1);
    struct ListNode *current = dummy;

    while (list1 != NULL && list2 != NULL) {
        if (list1->val < list2->val) {
            current->next = list1;
            list1 = list1->next;
        } else {
            current->next = list2;
            list2 = list2->next;
        }
        current = current->next;
    }

    current->next = list1 ? list1 : list2;

    struct ListNode *head = dummy->next;
    free(dummy);
    return head;
}

void printLinkedList(struct ListNode *head) {
    printf("[");
    while (head != NULL) {
        printf("%d", head->val);
        if (head->next != NULL)
            printf(" -> ");
        head = head->next;
    }
    printf("]\n");
}

void testMergeTwoLists(struct ListNode *list1, struct ListNode *list2) {
    printf("List 1: ");
    printLinkedList(list1);
    printf("List 2: ");
    printLinkedList(list2);

    struct timespec start, end;
    clock_gettime(CLOCK_MONOTONIC, &start);
    struct ListNode *result = mergeTwoLists(list1, list2);
    clock_gettime(CLOCK_MONOTONIC, &end);

    printf("Result: ");
    printLinkedList(result);

    double runtimeNanoSeconds = end.tv_nsec - start.tv_nsec;
    printf("Execution time: %.0lf ns\n", runtimeNanoSeconds);
}

int main(void) {
    struct ListNode *list1 = createNode(1);
    list1->next = createNode(2);
    list1->next->next = createNode(4);

    struct ListNode *list2 = createNode(1);
    list2->next = createNode(3);
    list2->next->next = createNode(4);

    testMergeTwoLists(list1, list2);

    return 0;
}
